test/bpf_tests/bpf_test.go

```diff
@@ -0,0 +1,470 @@
+			continue
+		}
+
+		loadAndRunSpec(t, entry, pinPath)
+	}
+}
+
+func loadAndRunSpec(t *testing.T, entry fs.DirEntry, pinPath string) {
+	elfPath := path.Join(*testPath, entry.Name())
+	spec := loadAndPrepSpec(t, elfPath)
+
+	const maxLoadRetries = 5
+	var (
+		coll    *ebpf.Collection
+		err     error
+		logSize = 1 << 20 // 1 MiB
+	)
+	for i := 0; i < maxLoadRetries; i++ {
+		coll, err = ebpf.NewCollectionWithOptions(spec, ebpf.CollectionOptions{
+			Programs: ebpf.ProgramOptions{
+				LogSize: logSize,
+			},
+			Maps: ebpf.MapOptions{
+				PinPath: pinPath,
+			},
+		})
+		if err != nil {
+			// Use this definition of ENOSPC instead of syscall.ENOSPC, since the stdlib constant is only
+			// available on linux. Even if this doesn't run, we still want it to compile on non-linux systems.
+			const ENOSPC = syscall.Errno(0x1c)
+			// ENOSPC usually means that the log size was to small, so increase and retry.
+			if errors.Is(err, ENOSPC) {
+				// 1, 4, 16, 64, 256 MiB
+				logSize = logSize << 2
+				continue
+			}
+
+			t.Fatal("new coll:", err)
+		}
+
+		break
+	}
+	defer coll.Close()
+
+	err = loadCallsMap(spec, coll)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	testNameToPrograms := make(map[string]programSet)
+
+	for progName, spec := range spec.Programs {
+		match := checkProgRegex.FindStringSubmatch(spec.SectionName)
+		if len(match) == 0 {
+			continue
+		}
+
+		progs := testNameToPrograms[match[1]]
+		if match[2] == "setup" {
+			progs.setupProg = coll.Programs[progName]
+		}
+		if match[2] == "check" {
+			progs.checkProg = coll.Programs[progName]
+		}
+		testNameToPrograms[match[1]] = progs
+	}
+
+	for progName, set := range testNameToPrograms {
+		if set.checkProg == nil {
+			t.Fatalf(
+				"File '%s' contains a setup program in section '%s' but no check program.",
+				elfPath,
+				spec.Programs[progName].SectionName,
+			)
+		}
+	}
+
+	// Collect debug events and add them as logs of the main test
+	var globalLogReader *perf.Reader
+	if coll.Maps["test_cilium_events"] != nil {
+		globalLogReader, err = perf.NewReader(coll.Maps["test_cilium_events"], 4096*16)
+		if err != nil {
+			t.Fatalf("new global log reader: %s", err.Error())
+		}
+		defer globalLogReader.Close()
+
+		linkCache := link.NewLinkCache()
+
+		go func() {
+			for {
+				rec, err := globalLogReader.Read()
+				if err != nil {
+					return
+				}
+
+				dm := monitor.DebugMsg{}
+				reader := bytes.NewReader(rec.RawSample)
+				if err := binary.Read(reader, byteorder.Native, &dm); err != nil {
+					return
+				}
+
+				t.Log(dm.Message(linkCache))
+			}
+		}()
+	}
+
+	for name, progs := range testNameToPrograms {
+		t.Run(name, subTest(progs, coll.Maps[suiteResultMap]))
+	}
+
+	if globalLogReader != nil {
+		// Give the global log buf some time to empty
+		// TODO: replace with flush on the buffer, as soon as cilium/ebpf supports that
+		time.Sleep(50 * time.Millisecond)
+	}
+}
+
+func loadAndPrepSpec(t *testing.T, elfPath string) *ebpf.CollectionSpec {
```

Comment:
(see my comment below on `cilium_xdp_scratch`. What about pulling in the first two commits of https://github.com/cilium/cilium/pull/19159 and building on top of those instead? It already includes the logic around prog types (although with a different heuristic) and MapSpec.Extra.

It also solves populating the calls map based on `X/Y` section names.

Reply:
I really wouldn't mind if we didn't complicating the PR any more than we have to. I would prefer to merge this and then iterate on it.

Reply:
Fine by me, as long as it gets done. :) It would eliminate much of the code in this file.